# sashimi.py

[See more example](https://sashimi.readthedocs.io/en/latest/)

## what is sashimi.py

sashimi.py is a tool for visualizing various next-generation sequencing (NGS) data, including DNA-seq, RNA-seq, single-cell RNA-seq and full-length sequencing datasets. 

### Features of sashimi.py

1. Support various file formats as input
2. Support strand-aware coverage plot
3. Visualize coverage by heatmap, including HiC diagram 
4. Visualize protein domain based given gene id
5. Demultiplex the single-cell RNA/ATAC-seq into cell population 
6. Support visualizing individual full-length reads in IGV-like way

## Input

sashimi.py supports almost NGS data format, including

- bam
- bed
- bigwig
- Depth file generated by `samtools depth`
- HiC

## Output

The output will be a pdf and other image file formats, and each track on output corresponds these datasets from config file.

## Usage

The sashimi.py is written in Python, and user could install it in a variety of ways as follows

1. install as command line tools

    ```bash
    git clone https://github.com/ygidtu/sashimi.py
    cd sashimi.py
    python setup.py install
    sashimipy --help
    ```
   
2. running from source

    ```bash
    git clone https://github.com/ygidtu/sashimi.py
    cd sashimi.py
    pip install -r requirements.txt
    python main.py --help
    ```
   
3. for `pipenv` users

    ```bash
    git clone https://github.com/ygidtu/sashimi.py
    cd sashimi.py
    pipenv install   # create virtualenv and install required packages
    pipenv shell   # switch to virtualenv
    python main.py --help
    ```

4. using docker image

    ```bash
    git clone https://github.com/ygidtu/sashimi.py
    cd sashimi.py
    docker build --rm -t ygidtu/sashimi.py .
    docker run --rm ygditu/sashimi.py --help
    ```


## Example

```bash
python main.py \
  -e chr1:1270656-1284730:+ \
  -r example/example.sorted.gtf.gz \
  --interval example/interval_list.tsv \
  --density example/density_list.tsv \
  --show-side \
  --igv example/igv.tsv \
  --heatmap example/heatmap_list.tsv \
  --focus 1272656-1272656:1275656-1277656 \
  --stroke 1275656-1277656:1277856-1278656@blue \
  --sites 1271656,1271656,1272656 \
  --line example/line_list.tsv \
  -o example/example.png \
  --dpi 300 \
  --width 10 \
  --height 1 \
  --barcode example/barcode_list.tsv \
  --domain --remove-duplicate-umi
```

![](example/example.png)

## Questions

Visit issues or contact Yiming Zhang(@ygidtu) and Ran Zhou(@zhou-ran)

## Citation

The citation will be available as soon as possible.
